File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int leftMax = 0, rightMax = 0, left = 0, right = n - 1, ans = 0;
        while (left < right) {
            if (height[left] <= height[right]) {
                if (height[left] > leftMax) leftMax = height[left];
                else ans += leftMax - height[left];
                left++;
            } else {
                if (height[right] > rightMax) rightMax = height[right];
                else ans += rightMax - height[right];
                right--;
            }
        }
        return ans;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int l = 0, r = height.size() - 1, lm = INT_MIN, rm = INT_MIN, res = 0;
        
        while (l <= r) {
            if (height[l] <= height[r]) {
                if (height[l] > lm) {
                    lm = height[l];
                } else {
                    res = res + (lm - height[l]);
                }
                l++;
            } else {
                if (height[r] > rm) {
                    rm = height[r];
                } else {
                    res = res + (rm - height[r]);
                }
                r--;
            }
        }
        
        return res;
    }
};